*: add bottlecap, an optional entry point

ajeddeloh opened the pull request:
Add bottlecap as the default entry point. It will handle launching the
coreos-assembler container and optionally binding in and rebuilding
certain aspects to aid development. This is in additon to the existing
flows.

Putting this up as WIP to get some feedback, will add docs before final version

cgwalters commented on bottlecap:
```diff
@@ -0,0 +1,65 @@
+		"--help")
+			usage
+			;;
+		*) # Rest of the args are for COSA
```
Optional: take an explicit `--` here too; this way one could e.g. run `coreos-assembler -- --help` and get the "inner help".

ajeddeloh replied:
fixed.

cgwalters commented on bottlecap:
```diff
@@ -0,0 +1,65 @@
+
+if [ "$dev" == 1 ]; then
+	volumes+="-v $script_dir:/host/src "
+	volumes+="-v $(go env GOCACHE):/root/.cache/go-build "
```
On a SELinux enabled system this assumes these are labeled too.  Which may be OK, but will almost certainly bite people.  (I try not to run containers accessing my `/home` for example)

ajeddeloh replied:
hmm. the gocache is easy enough (just put it literally anywhere that will persist across container runs).

If we want to support local dev though we need to be able to bind in wherever the build scripts are coming from (which is right now, this repo)

cgwalters replied:
The TL;DR is that SELinux + docker/podman (as root anyways) want files to be labeled specially for containers.  Trying to do `-v /home/myuser:/home` for example will fail, and you don't want to relabel your `/home`.

cgwalters replied:
(To pivot this topic though, while I haven't played with it the "rootless" podman should sidestep this issue; if we merge the bits to have this container not require privileges, just `/dev/kvm`, that will work well for rootless too)

dustymabe replied:
another note here: we need to check if `go env GOCACHE` reports a value before we try to mount it in:

```
$ ./bottlecap --build-dir /srv/ --dev --container ca build
./bottlecap: line 81: go: command not found
+ sudo podman run --rm --net=host -ti --privileged --userns=host -v /var/srv:/srv -v /var/sharedfolder/code/github.com/coreos/coreos-assembler:/host/src -v :/root/.cache/go-build --workdir /srv --entrypoint /host/src/bottlecap-shim ca d
invalid host path, must be an absolute path ""
```

cgwalters commented:
Can you elaborate slightly on the usage of this?  One would run it as root like `./bottlecap` from a `git clone` of this repo?

ajeddeloh commented:
Yeah, that's the idea. Basically allow an easy way to both work on the container itself and give an opinionated way to use it. Manually setting bash aliases is not ideal. The goal is also to keep `bottlecap` as stripped down as possible.

dustymabe commented on bottlecap:
```diff
@@ -0,0 +1,76 @@
+dev=0
+help=0
+runtime="podman"
+which podman > /dev/null 2> /dev/null || runtime="docker"
```
tip: `which podman &> /dev/null` sends both stdout and stderr to `/dev/null`

dustymabe commented on bottlecap:
```diff
@@ -0,0 +1,76 @@
+#!/bin/sh
+
+# bottlecap, 'cause it's kinda like cork... get it?
```
`cork` is nice since it's 4 letters :) - I like `cass` because it's also 4 letters, but I also like highlighting that this is the script vs an alias that someone has defined on the command line (since I like to define my alias as `alias cass=`. So `cass.sh` would be my suggestion here, but also takes us back to 7 characters so I'm a hypocrite :)

This was mostly just a thought excercise for me. We don't need to change anything unless you want to.

ajeddeloh commented:
Not sure what to do about the selinux issue @cgwalters brought up (ideas?) but I switched to getopt parsing.

dustymabe commented on bottlecap:
```diff
@@ -0,0 +1,80 @@
+	echo "  --container: which coreos-assembler container to use"
+	echo "  --help: print this help message"
+	echo "Command to be executed:"
+	runtime="echo $runtime"
```
I feel like the above two lines should be deleted.

ajeddeloh replied:
I could go either way. It's certainly been useful for debugging `bottlecap` and I think it does a good job of illustrating that `bottlecap` is just a thin wrapper. What about a `---dry-run` option instead?

dustymabe replied:
right now it doesn't actually output the command to be executed I don't think. It outputs `Command to be executed:` and then sets the `runtime` equal to `"echo $runtime"`, which doesn't print anything to the screen.

dustymabe replied:
I see. Thanks for explaining. I did not understand what you were doing. We discussed that you could also achieve this with `--dry-run`

dustymabe commented on bottlecap:
```diff
@@ -0,0 +1,80 @@
+	echo "usage:"
+	echo "bottlecap [--dev] [--runtime $runtime] [--build-dir $build_dir]  [--container $container] assemblerargs..."
+	echo "  --dev: rebuild and reinstall mantle, the build scripts, and the ostree-releng-scipts before running coreos-assembler"
+	echo "  --runtime: (auto) specify whether to use podman or docker. Defaults to podman if it is in \$PATH"
```
It might be better to say "Defaults to $runtime" and just let the default value from the script automatically be substituted in.

ajeddeloh replied:
Only concern there is someone looks at it on one system and thinks "Oh, the default is X", then writes some script for another system with that in mind and gets a different result.

dustymabe replied:
yeah, i could go either way. if we have to worry about people hacking on the script then all bets are off.

dustymabe commented on bottlecap:
```diff
@@ -0,0 +1,80 @@
+
+usage() {
+	echo "usage:"
+	echo "bottlecap [--dev] [--runtime $runtime] [--build-dir $build_dir]  [--container $container] assemblerargs..."
```
did you mean to evaluate the above variables when we print stuff to the screen? If not then we should use single quotes and not double quotes here.

ajeddeloh replied:
I do. You can do `bottlecap --build-dir foobar --help` and see your arg reflected in the help text, which I am a fan of personally.

dustymabe replied:
I see, I haven't seen that very much in practice, but seems OK

dustymabe commented on bottlecap:
```diff
@@ -0,0 +1,80 @@
+build_dir="$(pwd)"
+container="quay.io/cgwalters/coreos-assembler"
+
+TEMP=$(getopt -o 'dr:b:c:h' --long 'dev,runtime:,build-dir:,container:,help' -- $@)
```
you need to check the exit code of `getopt` and call usage() and exit if it failed.

dustymabe commented on bottlecap:
```diff
@@ -0,0 +1,80 @@
+	entrypoint="--entrypoint /host/src/bottlecap-shim"
+fi
+
+$runtime run --rm --net=host -ti --privileged --userns=host $volumes --workdir /srv $entrypoint $container $@
```
maybe we should `set -x` right above this command

also, do we expect users to run `sudo ./bottlecap` ?

ajeddeloh replied:
> also, do we expect users to run sudo ./bottlecap

If needed. I'd rather leave that to the user to decide than than bake it into the script.

ajeddeloh replied:
> maybe we should set -x right above this command

SGTM

dustymabe commented:
a few style comments:

- there is [a PR](#193) to run most of our bash through `shellcheck`, might be worth doing that here too.
- you like tabs don't you? I think we had that conversation before

dustymabe commented on bottlecap:
```diff
@@ -0,0 +1,80 @@
+done
+
+if [ "$help" == 1 ]; then
+	usage
```
we should exit 0 here?

ajeddeloh replied:
nope, see above comment about `runtime="echo $runtime"`

dustymabe replied:
I see, yep. I was did not understand what you were doing before.

ajeddeloh commented:
Will do on running it through shellcheck.

And yeah, I do like tabs, let everyone decide their own indentation level (I typically like 8, which is far more than most people I think).

ajeddeloh commented:
Fixed a lot of things, switch the gocache to just always use `$script_dir/.gocache` and added that to the `.gitignore`.

Passes shellcheck, with one exception where we actually do want glob expansion

dustymabe commented on bottlecap:
```diff
@@ -0,0 +1,103 @@
+rc=0
+TEMP=$(getopt -o 'dr:b:c:h' --long 'dev,dry-run,runtime:,build-dir:,container:,help' -- "$@") || rc=$?
+if [ "$rc" -ne 0 ]; then
+	echo "Bad arguments. getopt failed."
```
should the above go to stderr ?

dustymabe commented on bottlecap:
```diff
@@ -0,0 +1,103 @@
+dryrun=0
+help=0
+runtime="podman"
+command -v podman > /dev/null || runtime="docker"
```
what does the `command` command do? Did we switch from `which` for a reason?

dustymabe replied:
i have never seen that before.. nice. :+1:

ajeddeloh replied:
shellcheck complained about it

dustymabe replied:
interesting.. `command` is a builtin vs an external command so maybe that's why it's better

dustymabe commented on bottlecap:
```diff
@@ -0,0 +1,103 @@
+			break
+			;;
+		*)
+			echo "Error parsing args"
```
not sure if this section is needed any longer since we are using `getopt`, but if we keep this let's spit that message to stderr

dustymabe commented:
a few final comments in the code on stderr.. otherwise LGTM.

will probably want to update this in a future PR to handle the 'unpriv' case, but we don't need to do that now

ajeddeloh commented:
Made errors go to stdout, fixed shellcheck complaints in `bottlecap-shim`
